pinot-core/src/main/java/org/apache/pinot/core/operator/timeseries/TimeSeriesAggregationOperator.java

```diff
@@ -0,0 +1,251 @@
+    }
+    return timeIndexes;
+  }
+
+  public void processLongExpression(BlockValSet blockValSet, Map<Long, BaseTimeSeriesBuilder> seriesBuilderMap,
+      int[] timeValueIndexes, Object[][] tagValues) {
+    long[] valueColumnValues = blockValSet.getLongValuesSV();
+    for (int docIdIndex = 0; docIdIndex < timeValueIndexes.length; docIdIndex++) {
+      Object[] tagValuesForDoc = new Object[_groupByExpressions.size()];
+      for (int tagIndex = 0; tagIndex < tagValues.length; tagIndex++) {
+        tagValuesForDoc[tagIndex] = tagValues[tagIndex][docIdIndex];
+      }
+      long hash = TimeSeries.hash(tagValuesForDoc);
+      seriesBuilderMap.computeIfAbsent(hash,
+              k -> _seriesBuilderFactory.newTimeSeriesBuilder(_aggInfo, Long.toString(hash), _timeBuckets,
+                  _groupByExpressions,
+                  tagValuesForDoc))
+          .addValueAtIndex(timeValueIndexes[docIdIndex], (double) valueColumnValues[docIdIndex]);
+    }
+  }
+
+  public void processIntExpression(BlockValSet blockValSet, Map<Long, BaseTimeSeriesBuilder> seriesBuilderMap,
+      int[] timeValueIndexes, Object[][] tagValues) {
+    int[] valueColumnValues = blockValSet.getIntValuesSV();
+    for (int docIdIndex = 0; docIdIndex < timeValueIndexes.length; docIdIndex++) {
+      Object[] tagValuesForDoc = new Object[_groupByExpressions.size()];
+      for (int tagIndex = 0; tagIndex < tagValues.length; tagIndex++) {
+        tagValuesForDoc[tagIndex] = tagValues[tagIndex][docIdIndex];
+      }
+      long hash = TimeSeries.hash(tagValuesForDoc);
+      seriesBuilderMap.computeIfAbsent(hash,
+              k -> _seriesBuilderFactory.newTimeSeriesBuilder(_aggInfo, Long.toString(hash), _timeBuckets,
+                  _groupByExpressions,
+                  tagValuesForDoc))
+          .addValueAtIndex(timeValueIndexes[docIdIndex], (double) valueColumnValues[docIdIndex]);
+    }
+  }
+
+  public void processDoubleExpression(BlockValSet blockValSet, Map<Long, BaseTimeSeriesBuilder> seriesBuilderMap,
+      int[] timeValueIndexes, Object[][] tagValues) {
+    double[] valueColumnValues = blockValSet.getDoubleValuesSV();
+    for (int docIdIndex = 0; docIdIndex < timeValueIndexes.length; docIdIndex++) {
+      Object[] tagValuesForDoc = new Object[_groupByExpressions.size()];
+      for (int tagIndex = 0; tagIndex < tagValues.length; tagIndex++) {
+        tagValuesForDoc[tagIndex] = tagValues[tagIndex][docIdIndex];
+      }
+      long hash = TimeSeries.hash(tagValuesForDoc);
+      seriesBuilderMap.computeIfAbsent(hash,
+              k -> _seriesBuilderFactory.newTimeSeriesBuilder(_aggInfo, Long.toString(hash), _timeBuckets,
+                  _groupByExpressions,
+                  tagValuesForDoc))
+          .addValueAtIndex(timeValueIndexes[docIdIndex], valueColumnValues[docIdIndex]);
+    }
+  }
+
+  public void processStringExpression(BlockValSet blockValSet, Map<Long, BaseTimeSeriesBuilder> seriesBuilderMap,
```

Comment:
Will we have time series value as string?

Reply:
Yeah we could but only in the leaf stage. For instance, in M3 we would like to be able to support queries such as:

```
fetch table:foobar value:some_uuid_column
  | summarizeBy 1h countDistinct
  ...
```

This will essentially run a countDistinct in a 1 hour bucket on the `some_uuid_column`. This is a common use-case we have internally.

This will require an extension to the current TimeSeries code though. Right now we have tied TimeSeries to Double[] values. We will make it generic in a subsequent PR.